Add support for foreach commands ##newshell

ret2libc opened the pull request:
Adds support for all `@@@?` commands in the newshell.
This is mostly copy-paste code from existing `@@@?` code.

ret2libc commented:
By the way, this also fixes the issue in https://github.com/radareorg/radare2/issues/16600 . It does not fix the help msg though.

radare commented on libr/core/cmd.c:
```diff
@@ -4636,6 +4636,36 @@ DEFINE_IS_TS_FCN(double_quoted_arg)
+		}
+		char *cmt = (char *)sdb_decode (ptr + 1, 0);
+		if (cmt) {
+			r_core_cmdf (core, "s %s", k + 7);
```
use the API

XVilka replied:
Completely agree 👍

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+DEFINE_HANDLE_TS_FCN(foreach_addrsize_command) {
+	RCore *core = state->core;
+	TSNode command = ts_node_named_child (node, 0);
+	uint32_t i = 1;
```
use size_t

ret2libc replied:
tree-sitter uses `uint32_t` so I thought I should use it as well to avoid problems. `size_t` may be ut64 on some arch, while uint32_t is always 32bit.

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	RAnalBlock *bb;
+	bool ret = true;
+	r_list_foreach (fcn->bbs, iter, bb) {
+		r_core_seek (core, bb->addr, 1);
```
use true instead of 1, the arg expects bool

ret2libc replied:
good point... however I would feel a bit inconsistent in doing this here, as there are like ~250 uses of this function and probably ~200 use 0/1. Can I do a separate PR where I change ALL of them at once?

XVilka replied:
+1 for the separate PR.

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	TSNode in_cmd = ts_node_named_child (node, 1);
+	substitute_args_init (state, node);
+
+	RList *edits = r_list_newf ((RListFree)free_tsr2cmd_edit);
```
check for null

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	r_list_append (edits, e);
+
+	TSNode op = ts_node_child (node, 1);
+	e = create_cmd_edit (state, op, strdup ("@@@="));
```
passing strdup? isnt it leaking imho it should take a const char *

ret2libc replied:
It does not as `e` is then freed by `free_tsr2cmd_edit`, which expects the string to be free-able.

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	r_list_append (edits, e);
+
+	TSNode new_command;
+	bool ok = substitute_args_do (state, edits, &new_command);
```
unnecessary temporal constant variable which is read only once

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	TSNode command = ts_node_named_child (node, 0);
+	RBinImport *imp;
+	ut64 offorig = core->offset;
+	RList *list = r_bin_get_imports (core->bin);
```
can return null , better to do an early return in that case before allocating

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	r_list_foreach (list, iter, imp) {
+		char *impflag = r_str_newf ("sym.imp.%s", imp->name);
+		ut64 addr = r_num_math (core->num, impflag);
+		ut64 *n = R_NEW (ut64);
```
use ut64_new()

ret2libc replied:
that function is only defined/used in cmd_debug.c. I don't feel like it's that much used to make it generic at this point though.

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	r_list_foreach (lost, iter, naddr) {
+		ut64 addr = *naddr;
+		if (addr && addr != UT64_MAX) {
+			r_core_seek (core, addr, 1);
```
use true instead of 1

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	RList *list = r_bin_get_imports (core->bin);
+	RList *lost = r_list_newf (free);
+	RListIter *iter;
+	r_list_foreach (list, iter, imp) {
```
maybe better to just use RPVector or RVector

ret2libc replied:
sure, but... that's completely out of scope for this PR, which simply reuses existing code to implement existing commands in the newparser

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+		const char *item_name;
+		r_list_foreach (list, iter, item_name) {
+			value = r_reg_getv (core->dbg->reg, item_name);
+			r_core_seek (core, value, 1);
```
use true instead of 1

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	RBinSection *sec;
+	RListIter *iter;
+	r_list_foreach (obj->sections, iter, sec) {
+		r_core_seek (core, sec->vaddr, 1);
```
true vs 1

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+		RDebugMap *map;
+		RListIter *iter;
+		r_list_foreach (dbg->maps, iter, map) {
+			r_core_seek (core, map->addr, 1);
```
true vs 1

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+		r_core_seek (core, f->offset, 1);
+		res &= handle_ts_command (state, command);
+	}
+	r_core_seek (core, off, 0);
```
imho we should kill this 3rd argument to RCore.seek, but well.

radare replied:
but can be a bit tricky to use RIO curseek instead of RCore seek

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+		}
+		RListIter *iter;
+		r_list_foreach (list, iter, p) {
+			r_core_cmdf (core, "dp %d", p->pid);
```
use the api

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+		RListIter *iter;
+		r_list_foreach (list, iter, p) {
+			r_core_cmdf (core, "dp %d", p->pid);
+			r_cons_printf ("PID %d\n", p->pid);
```
unnecessary printf

ret2libc replied:
mmm.. I'm just mimicking existing commands and mostly copy-pasted existing code. It was already doing this printf and I copied it.

radare commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+			r_cons_printf ("PID %d\n", p->pid);
+			res &= handle_ts_command (state, command);
+		}
+		r_core_cmdf (core, "dp %d", origpid);
```
use the api

ret2libc commented on libr/core/cmd.c:
```diff
@@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
+	return res;
+}
+
+DEFINE_HANDLE_TS_FCN(foreach_thread_command) {
```
@radare I fail to see how this was working at all...
First, a semantic question: what does it mean to execute a command "foreach thread"?
Second: `dp %d` just executes `r_debug_pid_list`, which prints a bunch of PIDs on the screen, but does not really change anything (no seek, no block size, no nothing), so executing a command after this is just like executing a command X times (where X is the number of threads in the system).

XVilka replied:
For each thread of the debugging process. It's helpful to go through threads for diplaying the information or setting a breakpoints.

ret2libc replied:
Yes.. my point is that this has never worked :D

ret2libc replied:
I have fixed this in the newshell implementation to really work on all threads.

radare commented:
I know this and we ve discussed this already. Iterating with ut64 on 32 bit machines is a waste of time. Thats why loops should be using sizt for counters

> On 16 Apr 2020, at 13:42, Riccardo Schirone <notifications@github.com> wrote:
> 
> ﻿
> @ret2libc commented on this pull request.
> 
> In libr/core/cmd.c:
> 
> > @@ -5953,6 +5983,356 @@ DEFINE_HANDLE_TS_FCN(iter_hit_command) {
>  	return res;
>  }
>  
> +DEFINE_HANDLE_TS_FCN(foreach_addrsize_command) {
> +	RCore *core = state->core;
> +	TSNode command = ts_node_named_child (node, 0);
> +	uint32_t i = 1;
> tree-sitter uses uint32_t so I thought I should use it as well to avoid problems. size_t may be ut64 on some arch, while uint32_t is always 32bit.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

radare commented:
As long as its new code i dont see why not change it here

> On 16 Apr 2020, at 16:50, Riccardo Schirone <notifications@github.com> wrote:
> 
> ﻿
> @ret2libc commented on this pull request.
> 
> In libr/core/cmd.c:
> 
> > +
> +DEFINE_HANDLE_TS_FCN(foreach_bb_command) {
> +	RCore *core = state->core;
> +	TSNode command = ts_node_named_child (node, 0);
> +	RAnalFunction *fcn = r_anal_get_fcn_in (core->anal, core->offset, 0);
> +	ut64 offorig = core->offset;
> +	ut64 obs = core->blocksize;
> +	if (!fcn) {
> +		return false;
> +	}
> +
> +	RListIter *iter;
> +	RAnalBlock *bb;
> +	bool ret = true;
> +	r_list_foreach (fcn->bbs, iter, bb) {
> +		r_core_seek (core, bb->addr, 1);
> good point... however I would feel a bit inconsistent in doing this here, as there are like ~250 uses of this function and probably ~200 use 0/1. Can I do a separate PR where I change ALL of them at once?
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

radare commented:
I dont ser anything wrong in fixing. Opypasted code

> On 16 Apr 2020, at 16:57, Riccardo Schirone <notifications@github.com> wrote:
> 
> ﻿
> @ret2libc commented on this pull request.
> 
> In libr/core/cmd.c:
> 
> > +		.command = &command,
> +		.state = state,
> +	};
> +	r_meta_list_cb (state->core->anal, R_META_TYPE_COMMENT, 0, foreach_comment_newshell, (void *)&cmt_t, UT64_MAX);
> +	return true;
> +}
> +
> +DEFINE_HANDLE_TS_FCN(foreach_import_command) {
> +	RCore *core = state->core;
> +	TSNode command = ts_node_named_child (node, 0);
> +	RBinImport *imp;
> +	ut64 offorig = core->offset;
> +	RList *list = r_bin_get_imports (core->bin);
> +	RList *lost = r_list_newf (free);
> +	RListIter *iter;
> +	r_list_foreach (list, iter, imp) {
> sure, but... that's completely out of scope for this PR, which simply reuses existing code to implement existing commands in the newparser
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

radare commented:
Dbt @@@t will show the bscktrace of each thread. Which is selected on each itsration with dpt

> On 16 Apr 2020, at 17:05, Riccardo Schirone <notifications@github.com> wrote:
> 
> ﻿
> @ret2libc commented on this pull request.
> 
> In libr/core/cmd.c:
> 
> > +		if (r_cons_is_breaked ()) {
> +			break;
> +		}
> +		if (!filter || r_str_glob (fcn->name, filter)) {
> +			r_core_seek (core, fcn->addr, 1);
> +			r_core_block_size (core, r_anal_function_linear_size (fcn));
> +			res &= handle_ts_command (state, command);
> +		}
> +	}
> +	r_cons_break_pop ();
> +	r_core_block_size (core, obs);
> +	r_core_seek (core, offorig, 1);
> +	return res;
> +}
> +
> +DEFINE_HANDLE_TS_FCN(foreach_thread_command) {
> @radare I fail to see how this was working at all...
> First, a semantic question: what does it mean to execute a command "foreach thread"?
> Second: dp %d just executes r_debug_pid_list, which prints a bunch of PIDs on the screen, but does not really change anything (no seek, no block size, no nothing), so executing a command after this is just like executing a command X times (where X is the number of threads in the system).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

radare commented:
The threads are the ones in the child proxess u debug

> On 16 Apr 2020, at 17:05, Riccardo Schirone <notifications@github.com> wrote:
> 
> ﻿
> @ret2libc commented on this pull request.
> 
> In libr/core/cmd.c:
> 
> > +		if (r_cons_is_breaked ()) {
> +			break;
> +		}
> +		if (!filter || r_str_glob (fcn->name, filter)) {
> +			r_core_seek (core, fcn->addr, 1);
> +			r_core_block_size (core, r_anal_function_linear_size (fcn));
> +			res &= handle_ts_command (state, command);
> +		}
> +	}
> +	r_cons_break_pop ();
> +	r_core_block_size (core, obs);
> +	r_core_seek (core, offorig, 1);
> +	return res;
> +}
> +
> +DEFINE_HANDLE_TS_FCN(foreach_thread_command) {
> @radare I fail to see how this was working at all...
> First, a semantic question: what does it mean to execute a command "foreach thread"?
> Second: dp %d just executes r_debug_pid_list, which prints a bunch of PIDs on the screen, but does not really change anything (no seek, no block size, no nothing), so executing a command after this is just like executing a command X times (where X is the number of threads in the system).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ret2libc commented:
> Dbt @@@t will show the bscktrace of each thread. Which is selected on each itsration with dpt
> The threads are the ones in the child proxess u debug

I fail to see how it worked, as this is implemented in https://github.com/radareorg/radare2/blob/master/libr/core/cmd.c#L3923 and it uses `dp <pid>` (which just "List children of pid") and not `dp= <pid>`(which "Select pid").

ret2libc commented:
> I know this and we ve discussed this already. Iterating with ut64 on 32 bit machines is a waste of time. Thats why loops should be using sizt for counters

Indeed using a `size_t` when `uint32_t` is enough is a waste of time, why are you suggesting it? I'm using an external API that already uses `uint32_t`, I don't see why I should use `size_t` and then cast it to make tree-sitter accept it, when I can just directly use `uint32_t`.

radare commented:
not really. 64bt machines work better with 64bit registers

> On 16 Apr 2020, at 17:35, Riccardo Schirone <notifications@github.com> wrote:
> 
> 
> I know this and we ve discussed this already. Iterating with ut64 on 32 bit machines is a waste of time. Thats why loops should be using sizt for counters
> 
> Indeed using a size_t when uint32_t is enough is a waste of time, why are you suggesting it? I'm using an external API that already uses uint32_t, I don't see why I should use size_t and then cast it to make tree-sitter accept it, when I can just directly use uint32_t.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/radareorg/radare2/pull/16607#issuecomment-614727881>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAG75FTDVRE6GYMHSTMK4Q3RM4QUXANCNFSM4MJKVZXQ>.
>

radare commented:
so wha'ts wht you dont get from @@@t ?

ret2libc commented:
```
0x7f96ab0af33d]> pd 1 @@@t
 * 59084 ppid:59082 uid:102913 s /usr/local/bin/r2r
 - 59105 ppid:59084 uid:102913 S jq
PID 59084
            ;-- rip:
            0x7f96ab0af33d      c3             ret
PID 59104
            ;-- rip:
            0x7f96ab0af33d      c3             ret
 * 59084 ppid:59082 uid:102913 s /usr/local/bin/r2r
 - 59105 ppid:59084 uid:102913 S jq
```
As you can see, `pd 1` is executed two times on the same address, in the same process. The process is not changed.

radare commented:
then dp= is not executed? well i think the only problem here is missing ".dr*;sr PC" to reload the flags

> On 16 Apr 2020, at 19:37, Riccardo Schirone <notifications@github.com> wrote:
> 
> 
> 0x7f96ab0af33d]> pd 1 @@@t
>  * 59084 ppid:59082 uid:102913 s /usr/local/bin/r2r
>  - 59105 ppid:59084 uid:102913 S jq
> PID 59084
>             ;-- rip:
>             0x7f96ab0af33d      c3             ret
> PID 59104
>             ;-- rip:
>             0x7f96ab0af33d      c3             ret
>  * 59084 ppid:59082 uid:102913 s /usr/local/bin/r2r
>  - 59105 ppid:59084 uid:102913 S jq
> As you can see, pd 1 is executed two times on the same address, in the same process. The process is not changed.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/radareorg/radare2/pull/16607#issuecomment-614793966>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAG75FWZR2FVFXIIWURK6JDRM4645ANCNFSM4MJKVZXQ>.
>

XVilka commented:
Also related - it would be nice to have the textual command for `foreach` as well, in the future.

XVilka commented on libr/core/cmd.c:
```diff
@@ -4636,6 +4636,36 @@ DEFINE_IS_TS_FCN(double_quoted_arg)
+	struct tsr2cmd_state *state = cmt_t->state;
+	if (!strncmp (k, "meta.C.", 7)) {
+		char *ptr = strchr (v, ',');
+		if (!ptr || !*ptr) {
```
Don't we have a macro for this pattern? It's quite common.

ret2libc commented:
@XVilka @radare I have addressed the comments apart from:
- replacing `r_core_seek (...., 0/1)` with `r_core_seek (...., false/true)`. I'm going to do that in a separate PR, it's a completely separated problem from the purpose of this PR and it's useless to just fix this in 3 places and keep all the other 250 uses the same. I'm going to fix this once and for all separately.
- using size_t instead of uint32_t. I see no reason for doing this. The compiler will complain because the `i` is passed to a function that accepts `uint32_t`. Why shall we cast things around, when we can just use the right type?

Apart from that, I think the PR is good.
